python/cudf_polars/cudf_polars/experimental/shuffle.py

```diff
@@ -0,0 +1,245 @@
+    assume that the corresponding columns must be hashed. If
+    `keys` is a `NamedExpr`, we assume that the corresponding
+    column already contains a direct partition mapping.
+    """
+
+    __slots__ = ("keys", "options")
+    _non_child = ("schema", "keys", "options")
+    keys: tuple[NamedExpr, ...] | NamedExpr
+    """Keys to shuffle on."""
+    options: dict[str, Any]
+    """Shuffling options."""
+
+    def __init__(
+        self,
+        schema: Schema,
+        keys: tuple[NamedExpr, ...] | NamedExpr,
+        options: dict[str, Any],
+        *children: IR,
+    ):
+        self.schema = schema
+        self.keys = keys
+        self.options = options
+        self._non_child_args = ()
+        self.children = children
+        if len(children) > 2:  # pragma: no cover
+            raise ValueError(f"Expected a maximum of two children, got {children}")
+
+    def get_hashable(self) -> Hashable:
+        """Hashable representation of the node."""
+        return (
+            type(self),
+            tuple(self.schema.items()),
+            self.keys,
+            json.dumps(self.options),
+            self.children,
+        )
+
+    @classmethod
+    def do_evaluate(cls, df: DataFrame):  # pragma: no cover
+        """Evaluate and return a dataframe."""
+        # Single-partition logic is a no-op
+        return df
+
+
+def _partition_dataframe(
+    df: DataFrame,
+    index: DataFrame | None,
+    keys: tuple[NamedExpr, ...] | NamedExpr,
+    count: int,
+) -> dict[int, DataFrame]:
+    """
+    Partition an input DataFrame for shuffling.
+
+    Parameters
+    ----------
+    df
+        DataFrame to partition.
+    index
+        Optional DataFrame from which to extract partitioning
+        keys. If None, keys will be extracted from `df`.
+    keys
+        Shuffle key(s) to extract from index or df.
+    count
+        Total number of output partitions.
+
+    Returns
+    -------
+    A dictionary mapping between int partition indices and
+    DataFrame fragments.
+    """
+    # Extract output-partition mapping
+    if isinstance(keys, tuple):
+        # Hash the specified keys to calculate the output
+        # partition for each row (partition_map)
+        partition_map = plc.binaryop.binary_operation(
+            plc.hashing.murmurhash3_x86_32(
+                DataFrame([expr.evaluate(index or df) for expr in keys]).table
+            ),
+            plc.interop.from_arrow(pa.scalar(count, type="uint32")),
+            plc.binaryop.BinaryOperator.PYMOD,
+            plc.types.DataType(plc.types.TypeId.UINT32),
+        )
+    else:  # pragma: no cover
+        # Specified key column already contains the
+        # output-partition index in each row
+        partition_map = keys.evaluate(index or df).obj
+
+    # Apply partitioning
+    t, offsets = plc.partitioning.partition(
+        df.table,
+        partition_map,
+        count,
+    )
+
+    # Split and return the partitioned result
+    return {
```

Comment:
@madsbk - Just curious: When we integrate with `rapidsmp`, can we be able to simply call `pack` on the elements of this mapping and hand it off to the shuffle service?

Reply:
Yes, I think we can use [`partition_and_pack()`](https://github.com/rapidsai/rapids-multi-gpu/blob/83fae6c252bd4efe781219571416f72581ec288c/python/rapidsmp/rapidsmp/shuffler.pyx#L28)

Reply:
Okay cool - That sounds good for a hash-based shuffle, but we will not want hash partitioning for sorting. Rather, we will want to pass in a pylibcudf column containing the final partition for each row. Do we need to add that utility in rapidsmp?

Reply:
Yes, we would need an util function that takes a table and split+pack it into a `dict[PartID, PackedColumns]` based on the values in a column.